lib/SimpleSAML/Metadata/MetaDataStorageHandlerMDX.php

```diff
@@ -0,0 +1,287 @@
+		case 'shib13-sp-remote':
+			return $entity->getMetadata1xSP();
+		case 'attributeauthority-remote':
+			return $entity->getAttributeAuthorities();
+
+		default:
+			SimpleSAML_Logger::warning('MetaData - Handler.MDX: Unknown metadata set: ' . $set);
+		}
+
+		return NULL;
+	}
+
+
+	/**
+	 * Overriding this function from the superclass SimpleSAML_Metadata_MetaDataStorageSource.
+	 *
+	 * This function retrieves metadata for the given entity id in the given set of metadata.
+	 * It will return NULL if it is unable to locate the metadata.
+	 *
+	 * This class implements this function using the getMetadataSet-function. A subclass should
+	 * override this function if it doesn't implement the getMetadataSet function, or if the
+	 * implementation of getMetadataSet is slow.
+	 *
+	 * @param $index  The entityId or metaindex we are looking up.
+	 * @param $set  The set we are looking for metadata in.
+	 * @return An associative array with metadata for the given entity, or NULL if we are unable to
+	 *         locate the entity.
+	 */
+	public function getMetaData($index, $set) {
+		assert('is_string($index)');
+		assert('is_string($set)');
+
+		SimpleSAML_Logger::info('MetaData - Handler.MDX: Loading metadata entity [' . $index . '] from [' . $set . ']' );
+
+		/* Read from cache if possible. */
+		$data = $this->getFromCache($set, $index);
+
+		if ($data !== NULL && array_key_exists('expires', $data) && $data['expires'] < time()) {
+			/* Metadata has expired. */
+			$data = NULL;
+		}
+
+		if (isset($data)) {
+			/* Metadata found in cache and not expired. */
+			SimpleSAML_Logger::debug('MetaData - Handler.MDX: Using cached metadata for: ' . $index . '.');
+			return $data;
+		}
+
+		/* Look at Metadata Query Protocol: https://github.com/iay/md-query/blob/master/draft-young-md-query.txt */
+		$mdx_url = $this->server . '/entities/' . urlencode( $index);
+
+		SimpleSAML_Logger::debug('MetaData - Handler.MDX: Downloading metadata for "'. $index .'" from [' . $mdx_url . ']' );
+		try {
+			$xmldata = SimpleSAML_Utilities::fetch($mdx_url);
+		} catch(Exception $e) {
+			SimpleSAML_Logger::warning('Fetching metadata for ' . $index . ': ' . $e->getMessage());
+		}
+
+		if (empty($xmldata)) {
+			throw new Exception('Error downloading metadata for "'. $index .'" from "' . $mdx_url . '": ' .
+				SimpleSAML_Utilities::getLastError());
+		}
+
+		$entity = SimpleSAML_Metadata_SAMLParser::parseString($xmldata);
+		SimpleSAML_Logger::debug('MetaData - Handler.MDX: Completed parsing of [' .
+			$mdx_url . ']'  );
+
+
+		if( $this->validateFingerprint !== NULL) {
+			if(!$entity->validateFingerprint($this->validateFingerprint)) {
+				throw new Exception ('Error, could not verify signature for entity: ' . $index . '".');
+			}
+		}
+
+		$data = self::getParsedSet($entity, $set);
+		if ($data === NULL) {
+			throw new Exception('No metadata for set "' . $set .
+				'" available from "' . $index . '".');
+		}
+		if (is_array($data[0])) $data = $data[0];
```

Comment:
Ok, I fixed it.

The code `if (is_array($data[0])) $data = $data[0];` was needed because if `getParsedSet()` got `attributeauthority-remote`, then the metadata of the AttributeAuthority entity was under a '0' indexed subarray. If you look at `SAMLParser.php`, there is no function like `getMetadata20IdP` only `getAttributeAuthorities` which returns an array with subarrays. I didn't want to edit `SAMLParser.php`, so I handled this case at `getParsedSet()`.

I have tested, and it worked fine, without any warning or error message.